File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.py
"""
class RandomizedSet:

    def __init__(self):
        self.set = {}

    def insert(self, val: int) -> bool:
        not_present = False
        if self.set.get(val) == None:
            not_present = True
            self.set[val] = True
        return not_present
        

    def remove(self, val: int) -> bool:
        if self.set.get(val) == None:
            return False
        self.set.pop(val)
        return True
        
    def getRandom(self) -> int:
        return list(self.set.keys())[random.randint(0, len(self.set)-1)]
"""       

class RandomizedSet:

    def __init__(self):
        self.map = {}
        self.lst = []

    def insert(self, val: int) -> bool:
        if self.map.get(val) == None:
            self.map[val] = len(self.lst)
            self.lst.append(val)
            return True
        return False
        
    def remove(self, val: int) -> bool:
        if self.map.get(val) == None:
            return False
        
        idx = self.map[val]
        self.lst[idx] = self.lst[-1]
        self.map[self.lst[idx]] = idx
        
        self.lst.pop()
        self.map.pop(val)
        
        return True
        
    def getRandom(self) -> int:
        return self.lst[random.randint(0, len(self.lst)-1)]

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: 0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.go
import (
    "math/rand"
)

type RandomizedSet struct {
    m map[int]int
    s []int
}


func Constructor() RandomizedSet {
    return RandomizedSet{
        m: make(map[int]int),
        s: []int{},
    }
}


func (this *RandomizedSet) Insert(val int) bool {
    if _, ok := this.m[val]; ok {
        return false
    }
    this.m[val] = len(this.s)
    this.s = append(this.s, val)
    return true
}


func (this *RandomizedSet) Remove(val int) bool {
    if _, ok := this.m[val]; !ok {
        return false
    }
    idx := this.m[val]
    this.s[idx] = this.s[len(this.s)-1]
    this.m[this.s[idx]] = idx
    
    delete(this.m, val)
    this.s = this.s[:len(this.s)-1]
    return true
}


func (this *RandomizedSet) GetRandom() int {
    return this.s[rand.Intn(len(this.s))]
}


/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */